src/lib/root_prover/root_prover.ml

```diff
@@ -0,0 +1,160 @@
+    let open External_transition in
+    transition_with_hash |> With_hash.data |> Verified.protocol_state
+    |> Protocol_state.body |> Protocol_state.Body.hash
+
+  let consensus_state transition =
+    External_transition.(
+      protocol_state transition |> Protocol_state.consensus_state)
+
+  let prove ~frontier {logger; _} seen_consensus_state :
+      ( External_transition.t
+      , State_body_hash.t List.t * External_transition.t )
+      Proof_carrying_data.t
+      option =
+    let open Option.Let_syntax in
+    let best_tip_breadcrumb = Transition_frontier.best_tip frontier in
+    let best_tip =
+      Transition_frontier.Breadcrumb.transition_with_hash best_tip_breadcrumb
+      |> With_hash.data |> External_transition.of_verified
+    in
+    let is_tip_better =
+      Consensus.select ~logger ~existing:(consensus_state best_tip)
+        ~candidate:seen_consensus_state
+      = `Keep
+    in
+    let%bind () = Option.some_if is_tip_better () in
+    let exclusive_transitions_path =
+      Transition_frontier.path_map frontier best_tip_breadcrumb
+        ~f:(fun breadcrumb ->
+          let transition_with_hash =
+            Transition_frontier.Breadcrumb.transition_with_hash breadcrumb
+          in
+          Debug_assert.debug_assert (fun () ->
+              let state_hash = With_hash.hash transition_with_hash in
+              Logger.info logger
+                !"Hash of a breadcrumb in path: %{sexp:State_hash.t}"
+                state_hash ) ;
+          get_state_hash transition_with_hash )
+    in
+    let root_with_hash =
+      Transition_frontier.root frontier
+      |> Transition_frontier.Breadcrumb.transition_with_hash
+    in
+    Logger.info logger
+      !"Produced a merkle list of %{sexp:State_body_hash.t list}"
+      exclusive_transitions_path ;
+    Some
+      Proof_carrying_data.
+        { data=
+            root_with_hash |> With_hash.data |> External_transition.of_verified
+        ; proof= (exclusive_transitions_path, best_tip) }
+
+  let verify_merkle_proof ~logger root_state_hash merkle_list best_tip_hash =
+    let result_hash =
+      List.fold merkle_list ~init:root_state_hash ~f:(fun acc body_hash ->
+          let state_hash =
+            Protocol_state.hash ~hash_body:Fn.id
+              {previous_state_hash= acc; body= body_hash}
+          in
+          Logger.info logger
+            !"State_hash produce %{sexp:State_hash.t}"
+            state_hash ;
+          state_hash )
+    in
+    State_hash.equal best_tip_hash result_hash
+
+  let check_error ~message cond =
+    let open Deferred.Or_error in
+    if cond then return () else error_string message
+
+  let verify ~observed_state
+      ~peer_root:{Proof_carrying_data.data= root; proof= merkle_list, best_tip}
+      {logger; finality_length} :
+      ( External_transition.Proof_verified.t
+      * External_transition.Proof_verified.t )
+      Deferred.Or_error.t =
+    let open Deferred.Result.Let_syntax in
```

Comment:
Hashing is the single most expensive operation we do. You should hash as early as possible and maintain them for as long as possible. If the values this function takes in don't have precalculated hashes (which is seems like they don't since this comes over the network), this function should hash each of these transitions exactly once and return the transitions in a `With_hash.t` type to avoid hashing a second time outside the function.